File: Python/lld-examples/medium/online-shopping-system/shopping_cart.py
from typing import List, Dict, Optional
from product import Product

class CartItem:
    """Cart item class"""
    
    def __init__(self, product: Product, quantity: int):
        self.product = product
        self.quantity = quantity
    
    def get_total_price(self) -> float:
        """Get total price for this item"""
        return self.product.price * self.quantity
    
    def __str__(self):
        return f"{self.product.name} x{self.quantity} - ${self.get_total_price():.2f}"

class ShoppingCart:
    """Shopping cart class"""
    
    def __init__(self):
        self.items: List[CartItem] = []
        self.items_by_product: Dict[str, CartItem] = {}
    
    def add_item(self, product: Product, quantity: int) -> bool:
        """Add item to cart"""
        if not product.is_in_stock(quantity):
            return False
        
        if product.product_id in self.items_by_product:
            # Update existing item
            existing_item = self.items_by_product[product.product_id]
            new_quantity = existing_item.quantity + quantity
            
            if not product.is_in_stock(new_quantity):
                return False
            
            existing_item.quantity = new_quantity
        else:
            # Add new item
            cart_item = CartItem(product, quantity)
            self.items.append(cart_item)
            self.items_by_product[product.product_id] = cart_item
        
        return True
    
    def remove_item(self, product: Product):
        """Remove item from cart"""
        if product.product_id in self.items_by_product:
            cart_item = self.items_by_product[product.product_id]
            self.items.remove(cart_item)
            del self.items_by_product[product.product_id]
    
    def update_quantity(self, product: Product, quantity: int) -> bool:
        """Update item quantity"""
        if product.product_id in self.items_by_product:
            if quantity <= 0:
                self.remove_item(product)
                return True
            
            if not product.is_in_stock(quantity):
                return False
            
            self.items_by_product[product.product_id].quantity = quantity
            return True
        return False
    
    def clear_cart(self):
        """Clear cart"""
        self.items.clear()
        self.items_by_product.clear()
    
    def get_items(self) -> List[CartItem]:
        """Get all items in cart"""
        return self.items.copy()
    
    def get_total_price(self) -> float:
        """Get total price of cart"""
        return sum(item.get_total_price() for item in self.items)
    
    def get_item_count(self) -> int:
        """Get total number of items in cart"""
        return sum(item.quantity for item in self.items)
    
    def is_empty(self) -> bool:
        """Check if cart is empty"""
        return len(self.items) == 0
    
    def display_cart(self):
        """Display cart contents"""
        if self.is_empty():
            print("Cart is empty")
            return
        
        print("Cart contents:")
        for item in self.items:
            print(f"  {item}")
        print(f"Total items: {self.get_item_count()}")
        print(f"Total price: ${self.get_total_price():.2f}")
    
    def __str__(self):
        return f"Cart({self.get_item_count()} items, ${self.get_total_price():.2f})" 

File: JAVA/lld-examples/medium/online-shopping-system/src/main/java/com/systemdesign/lld/shopping/ShoppingCart.java
package com.systemdesign.lld.shopping;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * ShoppingCart manages cart items and operations
 */
public class ShoppingCart {
    private Map<String, CartItem> items;
    private String userId;
    
    public ShoppingCart(String userId) {
        this.userId = userId;
        this.items = new ConcurrentHashMap<>();
    }
    
    /**
     * Add item to cart
     */
    public boolean addItem(Product product, int quantity) {
        if (product == null || quantity <= 0) {
            return false;
        }
        
        // Check if enough stock is available
        if (!product.isAvailable(quantity)) {
            return false;
        }
        
        String productId = product.getProductId();
        CartItem existingItem = items.get(productId);
        
        if (existingItem != null) {
            int newQuantity = existingItem.getQuantity() + quantity;
            if (product.isAvailable(newQuantity)) {
                existingItem.updateQuantity(newQuantity);
                return true;
            }
            return false;
        } else {
            CartItem newItem = new CartItem(product, quantity);
            items.put(productId, newItem);
            return true;
        }
    }
    
    /**
     * Remove item from cart
     */
    public boolean removeItem(String productId) {
        return items.remove(productId) != null;
    }
    
    /**
     * Update item quantity
     */
    public boolean updateItemQuantity(String productId, int quantity) {
        CartItem item = items.get(productId);
        if (item != null) {
            if (quantity <= 0) {
                return removeItem(productId);
            }
            
            if (item.getProduct().isAvailable(quantity)) {
                item.updateQuantity(quantity);
                return true;
            }
        }
        return false;
    }
    
    /**
     * Get cart item by product ID
     */
    public CartItem getItem(String productId) {
        return items.get(productId);
    }
    
    /**
     * Get all cart items
     */
    public List<CartItem> getAllItems() {
        return new ArrayList<>(items.values());
    }
    
    /**
     * Get total price of all items in cart
     */
    public double getTotalPrice() {
        return items.values().stream()
                .mapToDouble(CartItem::getTotalPrice)
                .sum();
    }
    
    /**
     * Get total quantity of all items in cart
     */
    public int getTotalQuantity() {
        return items.values().stream()
                .mapToInt(CartItem::getQuantity)
                .sum();
    }
    
    /**
     * Check if cart is empty
     */
    public boolean isEmpty() {
        return items.isEmpty();
    }
    
    /**
     * Clear all items from cart
     */
    public void clearCart() {
        items.clear();
    }
    
    /**
     * Display cart contents
     */
    public void displayCart() {
        if (isEmpty()) {
            System.out.println("Cart is empty.");
            return;
        }
        
        System.out.println("Cart Contents:");
        for (CartItem item : items.values()) {
            Product product = item.getProduct();
            System.out.printf("- %s x%d - $%.2f each = $%.2f\n", 
                product.getName(), 
                item.getQuantity(), 
                product.getPrice(), 
                item.getTotalPrice());
        }
    }
    
    /**
     * Get cart size (number of different products)
     */
    public int getSize() {
        return items.size();
    }
    
    /**
     * Validate cart (check if all items are still available)
     */
    public boolean validateCart() {
        for (CartItem item : items.values()) {
            if (!item.getProduct().isAvailable(item.getQuantity())) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Get cart summary
     */
    public Map<String, Object> getCartSummary() {
        Map<String, Object> summary = new HashMap<>();
        summary.put("totalItems", getSize());
        summary.put("totalQuantity", getTotalQuantity());
        summary.put("totalPrice", getTotalPrice());
        summary.put("isEmpty", isEmpty());
        
        return summary;
    }
}

/**
 * CartItem represents an item in the shopping cart
 */
class CartItem {
    private Product product;
    private int quantity;
    
    public CartItem(Product product, int quantity) {
        this.product = product;
        this.quantity = quantity;
    }
    
    public Product getProduct() {
        return product;
    }
    
    public int getQuantity() {
        return quantity;
    }
    
    public void updateQuantity(int quantity) {
        this.quantity = quantity;
    }
    
    public double getTotalPrice() {
        return product.getPrice() * quantity;
    }
    
    @Override
    public String toString() {
        return "CartItem{" +
                "product=" + product.getName() +
                ", quantity=" + quantity +
                ", totalPrice=" + getTotalPrice() +
                '}';
    }
} 